src/Psalm/Type/Reconciler.php

```diff
@@ -96,6 +101,32 @@ public static function reconcileKeyedTypes(
             return $existing_types;
         }
 
+        $reference_map = [];
+        if (!empty($existing_references)) {
+            // PHP behaves oddly when passing an array containing references: https://bugs.php.net/bug.php?id=20993
```

Comment:
So [this](https://3v4l.org/3s7B5) is fine and makes perfect sense. But then you do [this](https://3v4l.org/G1VRc), which is the exact same thing but with an array item, and everything falls apart. So what you end up with is this [black magic](https://3v4l.org/tNYYj) to work around the problem.

I've come to the conclusion that references in PHP are a half-supported buggy mess, and if you use references in arrays or objects you're asking for trouble.